redash/serializers.py

```diff
@@ -45,9 +46,17 @@ def public_dashboard(dashboard):
     widget_list = (models.Widget.query
                    .filter(models.Widget.dashboard_id == dashboard.id)
                    .outerjoin(models.Visualization)
-                   .outerjoin(models.Query))
+                   .outerjoin(models.Query)
+                   #.with_entities(models.Widget, models.Query.latest_query_data_id)
+                   )
+
     widgets = {w.id: public_widget(w) for w in widget_list}
 
+    # make sure all widgets' query's have a last_query_data_id that is not null so public dashboards work
+    for w in widgets:
+        if not hasattr(w, 'latest_query_data'):
+            models.Query.query.filter(models.Query.id == widgets[w]['visualization']['query']['id'])
```

Comment:
I'm not exactly sure what you're saying. 

If you're saying this is rarely going to be called, you're right. It will mostly happen when people make dashboards via the API.

If you're saying there is no execution command (i.e. commit), could you please elaborate?